vm/actor/src/builtin/miner/deadline_assignment.rs

```diff
@@ -0,0 +1,185 @@
+    //    - If both have non-full partitions, assign to the most full one (stable assortment).
+    // 4. Assign to the deadline with the least number of live sectors.
+    // 5. Assign sectors to the deadline with the lowest index first.
+
+    // If one deadline would end up with fewer partitions (after
+    // compacting), assign to that one. This ensures we keep the maximum
+    // number of partitions in any given deadline to a minimum.
+    //
+    // Technically, this could increase the maximum number of partitions
+    // before compaction. However, that can only happen if the deadline in
+    // question could save an entire partition by compacting. At that point,
+    // the miner should compact the deadline.
+    a.compact_partitions_after_assignment(partition_size)
+        .cmp(&b.compact_partitions_after_assignment(partition_size))
+        .then_with(|| {
+            // If, after assignment, neither deadline would have fewer
+            // post-compaction partitions, assign to the deadline with the fewest
+            // pre-compaction partitions (after assignment). This will put off
+            // compaction as long as possible.
+            a.partitions_after_assignment(partition_size)
+                .cmp(&b.partitions_after_assignment(partition_size))
+        })
+        .then_with(|| {
+            // Ok, we'll end up with the same number of partitions any which way we
+            // go. Try to fill up a partition instead of opening a new one.
+            a.is_full_now(partition_size)
+                .cmp(&b.is_full_now(partition_size))
+        })
+        .then_with(|| {
+            // Either we have two open partitions, or neither deadline has an open
+            // partition.
+
+            // If we have two open partitions, fill the deadline with the most-full
+            // open partition. This helps us assign runs of sequential sectors into
+            // the same partition.
+            if !a.is_full_now(partition_size) && !b.is_full_now(partition_size) {
+                a.total_sectors.cmp(&b.total_sectors).reverse()
+            } else {
+                Ordering::Equal
+            }
+        })
+        .then_with(|| {
+            // Otherwise, assign to the deadline with the least live sectors. This
+            // will break the tie in one of the two immediately preceding
+            // conditions.
+            a.live_sectors.cmp(&b.live_sectors)
+        })
+        .then_with(|| {
+            // Finally, fall back on the deadline index.
+            a.index.cmp(&b.index)
+        })
+}
+
+// Assigns partitions to deadlines, first filling partial partitions, then
+// adding new partitions to deadlines with the fewest live sectors.
+pub fn assign_deadlines(
+    partition_size: u64,
+    deadlines: &[Option<Deadline>],
+    sectors: Vec<SectorOnChainInfo>,
+) -> Vec<Vec<SectorOnChainInfo>> {
+    struct Entry {
+        partition_size: u64,
```

Comment:
You're storing the partition size with every entry, is there a reason you do this instead of wrapping the `BinaryHeap` and just storing the partition size once? Is it just because of the trait bounds of the compares needed in the binary heap?

This isn't functionally a problem but depending on how many deadlines there are, it's just unnecessary duplicated data

Reply:
Yeah I'm not happy about this either, but `BinaryHeap` requires that the element type is `Ord` rather than being able to pass a custom comparator. So afaik there's literally no other way to do this, since `Ord::cmp` isn't allowed to access it otherwise